cli: add reset-quorum command

TheSamHuang opened the pull request:
This adds `reset-quorum` as a command in the debug CLI. This command utilizes the functionality added in #56333 to recover unavailable ranges where all replicas are lost.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57034)
<!-- Reviewable:end -->

TheSamHuang commented:
As we (@tbg, @irfansharif, @knz) discussed, the `UnsafeHealRange` RPC (introduced in #56333) only takes the node id and store id of the designated survivor as well as the range descriptor as inputs. Therefore, we must take the node id and store id as inputs to the command and we also need to get the range descriptor.

I'm currently unsure how to get the range descriptor (to update and write to meta ranges) and also would like some feedback on whether or not the `cli` package is the most appropriate place to do this. 

Any insight/feedback would be greatly appreciated -- thanks!

irfansharif commented:
> would like some feedback on whether or not the cli package is the most appropriate place to do this.

(I could be wrong, but) I was imagining the bulk of the heavy-lifting around reconstructing the appropriate range descriptors to happen on the server-side. Consider that in addition to #56333, we're also going to be introducing functionality to recover partially lost ranges where we'd need even more heavy lifting. I imagine it's more appropriate to happen within the server process itself. We'd want to keep the CLI UX minimal, asking the operator to possibly _only_ specify the range ID (I think), which would get sent to the server process via the RPC. At the server we'd use the primitive introduced in #56333, in addition to the future PR introducing quorum loss recovery, and tie it together after constructing the appropriate range descriptor. I don't think the CLI should be reading from KV directly.

TheSamHuang commented:
> I imagine it's more appropriate to happen within the server process itself. We'd want to keep the CLI UX minimal, asking the operator to possibly _only_ specify the range ID (I think), which would get sent to the server process via the RPC. 

I think @tbg envisioned the RPC taking in node ID, store ID, and the RangeDescriptor because the mechanism must be told what node to use as the designated survivor. Especially when recovering from an existing replica, we want to know where specifically we are recovering from.

TheSamHuang committed:
Adds reset-quorum CLI command. This command invokes the ResetQuorum RPC to
restore quorum to a given range ID. Note that data on any surviving replicas
will not be used to restore quorum. Instead, these replicas will be removed
irrevocably.

Release note: None.

irfansharif reviewed (commented):
Only looked at the last commit, which looks good! Be sure to change the commit message + title, and polish out the command help output. I'll rubber stamp it after the discussions in #56333 settle down and this is rebased atop that (we've also picked up a stray commit from Andrei here it seems).

tbg reviewed (commented):
Looks good!

> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-:-MN8AosqEeSK5axGs3Rx:b-vs1prh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @TheSamHuang)

___
*[pkg/cli/debug_reset_quorum.go, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN0BxwMsWugtmxumFB_-r3-22:-MN84hjs-BQprTPHCgIv:bi0jxib) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L22)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Reminder to drop the `--survivor-store-id` flag and adjust to `cobra.ExactArgs(1)` below.
</blockquote></details>

Does it make sense to use a flag for the rangeID? It is required, so why isn't it just the first parameter?
BTW as written we need `cobra.NoArgs`, though with the rangeID turning into a parameter it would become `cobra.ExactArgs(1)` as Irfan suggested.

Ah, I am seeing now that the `Use` line is just wrong and the impl is already using positional args, good. But let's fix the `Use` line :-)
___
*[pkg/cli/debug_reset_quorum.go, line 23 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN853FP1b1tWtZBH06c:-MN853FP1b1tWtZBH06d:bfwf5e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L23)):*
> ```Go
> var debugResetQuorumCmd = &cobra.Command{
> 	Use:   "reset-quorum --survivor-store-id=[store ID] --range-id=[range ID]",
> 	Short: "Reset quorum on the given range by designating the current node as the sole voter.",
> ```

target node
___
*[pkg/cli/debug_reset_quorum.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN858I930wqKv7jJZK6:-MN858I930wqKv7jJZK7:b-dldvvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L29)):*
> ```Go
> 
> This command is UNSAFE and should only be used with the supervision of
> a Cockroach Labs engineer. It is a last-resort option to recover a specified 
> ```

supervision of Cockroach Labs support.
___
*[pkg/cli/debug_reset_quorum.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN85-5QEPbqKQGi2NAM:-MN85-5QEPbqKQGi2NAN:b-jlwake) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L31)):*
This is already pretty good. I would add:

> Data on any surviving replicas will *not* be used to restore quorum. Instead, these replicas will be removed irrevocably.
___
*[pkg/cli/debug_reset_quorum.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8APwq7XWSI63HKPkY:-MN8APwq7XWSI63HKPkZ:b-drw7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L37)):*
> ```Go
> }
> 
> func runDebugResetQuorumRequest(cmd *cobra.Command, args []string) error {
> ```

`runDebugResetQuorum`
___
*[pkg/cli/debug_reset_quorum.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8AdPt67hZCOCHivk8:-MN8AdPt67hZCOCHivk9:b4p1beb) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L41)):*
> ```Go
> 	defer cancel()
> 
> 	storeID, err := strconv.ParseInt(args[0], 10, 32)
> ```

This goes away.
___
*[pkg/cli/debug_reset_quorum.go, line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8Abaq3d3eslcf9r-a:-MN8Abaq3d3eslcf9r-b:bn8btba) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L46)):*
> ```Go
> 	}
> 
> 	rangeID, err := strconv.ParseInt(args[1], 10, 32)
> ```

args[0]
___
*[pkg/cli/debug_reset_quorum.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8Aglq1iHgWF6IdqYF:-MN8Aglq1iHgWF6IdqYG:b8x6xt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L58)):*
> ```Go
> 	}
> 	defer finish()
> 	// Call ResetQuorum to reset quorum for given range on current node.
> ```

target node rather than current node


<!-- Sent from Reviewable.io -->

TheSamHuang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, @tbg, and @TheSamHuang)

___
*[pkg/cli/debug_reset_quorum.go, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN0BxwMsWugtmxumFB_-r3-22:-MN_4Sm97EmQhnaH1PAz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L22)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Does it make sense to use a flag for the rangeID? It is required, so why isn't it just the first parameter?
BTW as written we need `cobra.NoArgs`, though with the rangeID turning into a parameter it would become `cobra.ExactArgs(1)` as Irfan suggested.

Ah, I am seeing now that the `Use` line is just wrong and the impl is already using positional args, good. But let's fix the `Use` line :-)
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 23 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN853FP1b1tWtZBH06c:-MN_4f1L0hhf2_na1ceQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

target node
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN858I930wqKv7jJZK6:-MN_4e5R-waW0QHoXLcL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L29)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

supervision of Cockroach Labs support.
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN85-5QEPbqKQGi2NAM:-MN_4hrC9VohX-Ohzf-f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L31)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is already pretty good. I would add:

> Data on any surviving replicas will *not* be used to restore quorum. Instead, these replicas will be removed irrevocably.

</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8APwq7XWSI63HKPkY:-MN_4kA6AhV2mXuDFhk6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L37)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`runDebugResetQuorum`
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8AdPt67hZCOCHivk8:-MN_4lU46-3OvCA_Sjp2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L41)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This goes away.
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN0BxwMsWugtmxumFB_-r3-53:-MN_4lh8Ced5aQuZwgEL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L53)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Unnecessary line here.
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN0BxwMsWugtmxumFB_-r3-58:-MN_4xTCBkIgwqGhpYkm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L58)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Add new line above.
</blockquote></details>

Done.
___
*[pkg/cli/debug_reset_quorum.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8Aglq1iHgWF6IdqYF:-MN_4xk85HlRwGdK3bpt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L58)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

target node rather than current node
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

TheSamHuang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/cli/debug_reset_quorum.go, line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MN8Abaq3d3eslcf9r-a:-MN_A5j86rxPv9yrfI0_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/387d8208b337f9f548d60a8fbe133269bf1ae23f/pkg/cli/debug_reset_quorum.go#L46)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

args[0]
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @tbg, and @TheSamHuang)

___
*[pkg/cli/debug_reset_quorum.go, line 64 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MNcyJPi-VYVLontosvc:-MNcyJPi-VYVLontosvd:bxe3kl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8c484581176c6b4424c0afca8b315d36412854/pkg/cli/debug_reset_quorum.go#L64)):*
> ```Go
> 		return err
> 	}
> 	fmt.Printf("ok; please verify https://<ui>/#/reports/range/%d", rangeID)
> ```

This misses a final newline.


<!-- Sent from Reviewable.io -->

TheSamHuang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/cli/debug_reset_quorum.go, line 64 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57034#-MNcyJPi-VYVLontosvc:-MNd01ED5uFhsh-6vIxd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8c484581176c6b4424c0afca8b315d36412854/pkg/cli/debug_reset_quorum.go#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This misses a final newline.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 1 of 2 files at r3, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57034)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)



<!-- Sent from Reviewable.io -->

TheSamHuang commented:
bors r=knz
